zcash_client_sqlite/src/wallet/init/migrations/full_account_ids.rs

```diff
@@ -0,0 +1,381 @@
+use std::{collections::HashSet, rc::Rc};
+
+use rusqlite::{params, Transaction};
+use schemer_rusqlite::RusqliteMigration;
+use secrecy::SecretVec;
+use uuid::Uuid;
+use zcash_keys::keys::HdSeedFingerprint;
+
+use crate::wallet::init::WalletMigrationError;
+
+use super::{add_account_birthdays, receiving_key_scopes, v_transactions_note_uniqueness};
+
+/// The migration that switched from presumed seed-derived account IDs to supporting
+/// HD accounts and all sorts of imported keys.
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x1b104345_f27e_42da_a9e3_1de22694da43);
+
+pub(crate) struct Migration {
+    pub(super) seed: Rc<Option<SecretVec<u8>>>,
+}
+
+impl schemer::Migration for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [
+            receiving_key_scopes::MIGRATION_ID,
+            add_account_birthdays::MIGRATION_ID,
+            v_transactions_note_uniqueness::MIGRATION_ID,
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Replaces the `account` column in the `accounts` table with columns to support all kinds of account and key types."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            r#"
+            PRAGMA foreign_keys = OFF;
+            PRAGMA legacy_alter_table = ON;
+
+            CREATE TABLE accounts_new (
+                id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                account_type INTEGER NOT NULL DEFAULT 0,
+                hd_seed_fingerprint BLOB,
+                hd_account_index INTEGER,
+                uvk TEXT NOT NULL,
+                birthday_height INTEGER NOT NULL,
+                recover_until_height INTEGER,
+                CHECK (
+                    (account_type = 0 AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL)
+                    OR
+                    (account_type = 1 AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL)
+                )
+            );
+            CREATE UNIQUE INDEX accounts_uvk ON accounts_new ("uvk");
+            "#,
+        )?;
```

Comment:
The `CHECK` here cannot be changed later without doing another "create new table, copy all the data across, delete old table, rename new table" cycle. Given that the only other account types I'm aware of that might be added are non-ZIP-32 types, we should pre-emptively just gate the other half of this check on `account_type != 0`.

Also, avoid the magic numbers by deriving the constants from the enums.
```suggestion
        let account_type_zip32 = u32::from(AccountType::Zip32);
        transaction.execute_batch(&format!(
            r#"
            PRAGMA foreign_keys = OFF;
            PRAGMA legacy_alter_table = ON;

            CREATE TABLE accounts_new (
                id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
                account_type INTEGER NOT NULL DEFAULT 0,
                hd_seed_fingerprint BLOB,
                hd_account_index INTEGER,
                uvk TEXT NOT NULL,
                birthday_height INTEGER NOT NULL,
                recover_until_height INTEGER,
                CHECK (
                    (account_type = {account_type_zip32} AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL)
                    OR
                    (account_type != {account_type_zip32} AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL)
                )
            );
            CREATE UNIQUE INDEX accounts_uvk ON accounts_new ("uvk");
            "#,
        ))?;
```

Reply:
I like the != 0 idea.

I disagree on the magic numbers. We don't want the behavior of this migration to subtly change over time if the enum's constant value is changed. A change in those numbers would merit another migration to update the db schema. It's similar to why I had to copy some product code into migrations in order to freeze their sql queries to what works at the time those migrations execute.

Reply:
It's going to be necessary to keep those magic numbers stable and consistent with the values of the enum, so I agree with @str4d here; there should be one place where we define the canonical mapping from `AccountType` to integer.

I've done something similar with `PoolType`, although for that rather than defining a `From` instance I used an explicit function (I'm generally suspicious of using `From` for encoding/decoding, because they're general type conversion traits, not use-specific encoding and decoding.)

Reply:
The fact that I made the suggestion here is mainly because I saw it here first; I could have equally made it in `init.rs` where we pin the current database state, and that absolutely needs to exactly match the enum variant integer encodings at all times. As I said in my earlier comment, I don't mind if one of the migration or the pin uses magic numbers, as long as one of them does not.